NuGet restore on solution with Visual Studio Installer projects (vdproj) causes MSB4025

Solutions with legacy Visual Studio installer projects ( vdproj)  have started failing.

It would seem it doesn't result in a non zero exit code, but also doesn't restore all project properly as we MSBuild later fails with 'project.assets.json' not found. Run a NuGet package restore to generate this file.

## Details about Problem

NuGet product used: NuGet.exe

NuGet version (x.x.x.xxx): 4.9.3

Worked before? If so, with which NuGet version: 4.7.1

## Detailed repro steps so we can see the same problem

1. NuGet.exe restore "TestSolution.sln"

### Verbose Logs

TestProjectInstaller\TestProjectInstaller.vdproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, 
position 1.

### Sample Project

Commuting right now, but will see if I can get a minimal repo project.

But the project causing issues has
1. VDPROJ installer
2. A multitargeted net462 & netstandard1.5 class lib
3. v4.5.2 console app

Repro attached [TestSolutionA.zip](https://github.com/NuGet/Home/files/2866084/TestSolutionA.zip)

## Log running 4.7.1 - ✔️ 
```
λ .\v4.7.1_nuget.exe restore .\TestSolutionA\TestSolution.sln
MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin'.
Restoring NuGet package Newtonsoft.Json.12.0.1.
Adding package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'
Added package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'
Restoring packages for C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\MultiTargetLibrary.csproj...
  GET https://api.nuget.org/v3-flatcontainer/system.net.http/index.json
  OK https://api.nuget.org/v3-flatcontainer/system.net.http/index.json 617ms
  GET https://api.nuget.org/v3-flatcontainer/system.net.http/4.3.3/system.net.http.4.3.3.nupkg
  OK https://api.nuget.org/v3-flatcontainer/system.net.http/4.3.3/system.net.http.4.3.3.nupkg 16ms
Installing System.Net.Http 4.3.3.
Committing restore...
Generating MSBuild file C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\obj\MultiTargetLibrary.csproj.nuget.g.props.
Generating MSBuild file C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\obj\MultiTargetLibrary.csproj.nuget.g.targets.
Writing lock file to disk. Path: C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\obj\project.assets.json
Restore completed in 5,01 sec for C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\MultiTargetLibrary.csproj.

NuGet Config files used:
    C:\Users\MattiasKarlsson\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    C:\Users\MattiasKarlsson\.nuget\packages\
    https://api.nuget.org/v3/index.json
    C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\

Installed:
    1 package(s) to packages.config projects
    1 package(s) to C:\temp\priv\nuget\repro\TestSolutionA\MultiTargetLibrary\MultiTargetLibrary.csproj
```

## Log running 4.8.1 ❌ 

```
 λ .\v4.8.1_nuget.exe restore .\TestSolutionA\TestSolution.sln
MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin'.
C:\temp\priv\nuget\repro\TestSolutionA\TestProjectInstaller\TestProjectInstaller.vdproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.

WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
Restoring NuGet package Newtonsoft.Json.12.0.1.
Adding package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'
Added package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'

NuGet Config files used:
    C:\Users\MattiasKarlsson\AppData\Roaming\NuGet\NuGet.Config
    C:\Program Files (x86)\NuGet\Config\Microsoft.VisualStudio.Offline.config

Feeds used:
    C:\Users\MattiasKarlsson\.nuget\packages\
    https://api.nuget.org/v3/index.json
    C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\

Installed:
    1 package(s) to packages.config projects
```

## Log running 4.9.3 ❌ 

```
 λ .\v4.9.3_nuget.exe restore .\TestSolutionA\TestSolution.sln -Verbosity Detailed
NuGet Version: 4.9.3.5777
MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
MSBuild P2P timeout [ms]: 120000
C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\msbuild.exe "C:\Users\MattiasKarlsson\AppData\Local\Temp\NuGetScratch\ifozcrwc.u1i.nugetinputs.targets" /t:GenerateRestoreGraphFile /nologo /nr:false /v:q /p:NuGetRestoreTargets="C:\Users\MattiasKarlsson\AppData\Local\Temp\NuGetScratch\r21mghmk.g1z.nugetrestore.targets" /p:RestoreUseCustomAfterTargets="True" /p:RestoreTaskAssemblyFile="C:\temp\priv\nuget\repro\v4.9.3_nuget.exe" /p:RestoreSolutionDirectory="C:\temp\priv\nuget\repro\TestSolutionA\\" /p:SolutionDir="C:\temp\priv\nuget\repro\TestSolutionA\\" /p:SolutionName="TestSolution"
C:\temp\priv\nuget\repro\TestSolutionA\TestProjectInstaller\TestProjectInstaller.vdproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.

NuGet.CommandLine.ExitCodeException: Exception of type 'NuGet.CommandLine.ExitCodeException' was thrown.
   at NuGet.CommandLine.MsBuildUtility.<GetProjectReferencesAsync>d__6.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<GetDependencyGraphSpecAsync>d__52.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NuGet.CommandLine.RestoreCommand.<DetermineInputsFromMSBuildAsync>d__47.MoveNext()
WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
Restoring NuGet package Newtonsoft.Json.12.0.1.
Missing C:\Program Files (x86)\Microsoft SDKs\NuGetPackages\newtonsoft.json\12.0.1\newtonsoft.json.12.0.1.nupkg
Adding package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'
PackageSignatureVerificationLog: PackageIdentity: Newtonsoft.Json.12.0.1 Source: C:\Users\MattiasKarlsson\.nuget\packages\ PackageSignatureValidity: True
Added package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages'
Added package 'Newtonsoft.Json.12.0.1' to folder 'C:\temp\priv\nuget\repro\TestSolutionA\packages' from source 'C:\Users\MattiasKarlsson\.nuget\packages\'
```

Interestingly even though it's an failure exit code is still `0`

@nkolev92 yes if I for the above attached repro solution execute 
```powershell
$env:RestoreUseSkipNonexistentTargets='false'
nuget restore .\TestSolution.sln
```
nuget.exe doesn't produce any warnings and following MSBuild succeeds 👍  , so that sorts the issue, but still feels like a regression.



> The reason for the non-zero exit code also seems to be deliberate as the warning indicates:
> 
> > WARNING: Error reading msbuild project information, ensure that your input solution or project file is > valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.

Sure but if you see log above I also get an error

```diff
- TestProjectInstaller.vdproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
- NuGet.CommandLine.ExitCodeException: Exception of type 'NuGet.CommandLine.ExitCodeException' was thrown.
-    at NuGet.CommandLine.MsBuildUtility.<GetProjectReferencesAsync>d__6.MoveNext()
- --- End of stack trace from previous location where exception was thrown ---
-    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
-    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
-    at NuGet.CommandLine.RestoreCommand.<GetDependencyGraphSpecAsync>d__52.MoveNext()
- --- End of stack trace from previous location where exception was thrown ---
-    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
-    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
-    at NuGet.CommandLine.RestoreCommand.<DetermineInputsFromMSBuildAsync>d__47.MoveNext()
```

Also is there a way to treat warnings as errors? Setting `TreatWarningsAsErrors=true` didn't have any affect. A failed restore just means it'll fail later in the build process, I rather when possible fail earlier and not waste build minutes.

Any update? I have a same issue when building with AppVeyor

```
Build started
git clone -q --branch=develop https://github.com/d2phap/ImageGlass.git C:\projects\imageglass
git checkout -qf 8e0a114f45f70af6fbb8085c028da3f3ca5938da
cd Source
"ImageGlass/resources.bat"
nuget restore
ImageGlass/resources.bat
MSBuild auto-detection: using msbuild version '16.0.360.33175' from 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\bin'.
nuget : C:\projects\imageglass\Source\Setup\MSI\ImageGlass\ImageGlass.vdproj(1,1): error MSB4025: The project file could not be loaded. Data at the root level is invalid. Line 1, position 1.
At line:3 char:1
+ nuget restore
+ ~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (C:\projects\ima... 1, position 1.:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
```

@nkolev92  - this is also apparently ignoring if a project is included in the build configuration or not for a solution. I've got a pair of dtproj files in a solution that are not expected to be built in addition to a web site project thats also should not be getting built, but nuget restore task craps on reading the proj file  and doesnt restore and create the project.assets,.json for projects that are in the solution that didn't have any errors.

```
2019-08-07T16:23:55.3675752Z [command]C:\hostedtoolcache\windows\NuGet\4.9.4\x64\nuget.exe restore D:\a\3\s\Program1\.sln -Verbosity Detailed -NonInteractive -ConfigFile D:\a\3\Nuget\tempNuGet_9201.config
2019-08-07T16:24:25.0813025Z D:\a\3\s\Program1 Integration Services\Program1\ Integration Services.dtproj(3,3): error MSB4067: The element <DeploymentModel> beneath element <Project> is unrecognized.
2019-08-07T16:24:25.0813246Z D:\a\3\s\Program2 Integration Services\Program2 Integration Services.dtproj(3,3): error MSB4067: The element <DeploymentModel> beneath element <Project> is unrecognized.
2019-08-07T16:24:25.0813277Z 
2019-08-07T16:24:25.0813344Z NuGet Version: 4.9.4.5839
2019-08-07T16:24:25.0813397Z MSBuild auto-detection: using msbuild version '15.9.21.664' from 'C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin'. Use option -MSBuildVersion to force nuget to use a specific version of MSBuild.
2019-08-07T16:24:25.0813481Z WARNING: Project file D:\a\3\s\Program1\DBApp\client cannot be found.
2019-08-07T16:24:25.0813519Z MSBuild P2P timeout [ms]: 120000
2019-08-07T16:24:25.0813603Z C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\bin\msbuild.exe "C:\Users\VssAdministrator\AppData\Local\Temp\NuGetScratch\vlwntotk.vyj.nugetinputs.targets" /t:GenerateRestoreGraphFile /nologo /nr:false /v:q /p:NuGetRestoreTargets="C:\Users\VssAdministrator\AppData\Local\Temp\NuGetScratch\ktckdjxx.vgm.nugetrestore.targets" /p:RestoreUseCustomAfterTargets="True" /p:RestoreTaskAssemblyFile="C:\hostedtoolcache\windows\NuGet\4.9.4\x64\nuget.exe" /p:RestoreSolutionDirectory="D:\a\3\s\\" /p:RestoreConfigFile="D:\a\3\Nuget\tempNuGet_9201.config" /p:SolutionDir="D:\a\3\s\\" /p:SolutionName="Program1\"
2019-08-07T16:24:25.0813667Z NuGet.CommandLine.ExitCodeException: Exception of type 'NuGet.CommandLine.ExitCodeException' was thrown.
2019-08-07T16:24:25.0813717Z    at NuGet.CommandLine.MsBuildUtility.<GetProjectReferencesAsync>d__6.MoveNext()
2019-08-07T16:24:25.0813777Z --- End of stack trace from previous location where exception was thrown ---
2019-08-07T16:24:25.0813815Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2019-08-07T16:24:25.0813856Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2019-08-07T16:24:25.0813914Z    at NuGet.CommandLine.RestoreCommand.<GetDependencyGraphSpecAsync>d__52.MoveNext()
2019-08-07T16:24:25.0813953Z --- End of stack trace from previous location where exception was thrown ---
2019-08-07T16:24:25.0813990Z    at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
2019-08-07T16:24:25.0814212Z    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
2019-08-07T16:24:25.0814273Z    at NuGet.CommandLine.RestoreCommand.<DetermineInputsFromMSBuildAsync>d__47.MoveNext()
2019-08-07T16:24:25.0814551Z WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
2019-08-07T16:24:25.0814592Z Restoring NuGet package EntityFramework.6.1.3.
```

From the data below, something changed in the MsBuild after `15.9*` versions in the way how it handles `RestoreBuildInParallel, RestoreUseSkipNonexistentTargets, ContinueOnError ` properties.

nuget.exe   version | msbuild.exe version | error raised | Additional environment   variables | restore result
-- | -- | -- | -- | --
4.7.3 | 15.9.21.664 | MsBuild raised errors but   returned 0 exit code and hence the restore succeeded.            \TestSolutionA-4732017\TestSolutionA\TestProjectInstaller\TestProjectInstaller.vdproj(1,1):   error MSB4025: The project file could not be loaded. Data at the root level   is invalid. Line 1, position 1.       C:\Users\kapenaga\Downloads\TestSolutionA-4732017\TestSolutionA\TestProjectInstaller\TestProjectInstaller.vdproj   : warning NU1503: Skipping restore for project   'C:\Users\kapenaga\Downloads\TestSolutionA-4732017\TestSolutionA\TestProjectInstaller\TestProjectInstaller.vdproj'.   The project file may be invalid or missing targets required for restore. | RestoreBuildInParallel and        RestoreUseSkipNonexistentTargets properties set to false through code.   ContinueOnError set to "WarnAndContinue" in NuGet.targets file | Success
4.7.3 | 16.11.2.50704 or 17.2.0.16512 | MsBuild raised error and   returned non-zero exit code and hence restore failed with an exception. | RestoreBuildInParallel and RestoreUseSkipNonexistentTargets properties set to false through code.   ContinueOnError set to "WarnAndContinue" in NuGet.targets file | Failed
6.1.0 | 15.9.21.664 or 16.11.2.50704 or 17.2.0.16512 | MsBuild raised error and   returned non-zero exit code and hence restore failed with an exception. | RestoreBuildInParallel and        RestoreUseSkipNonexistentTargets properties set to true in NuGet.targets   file | Failed

NuGet 4.7.3 version passes `RestoreBuildInParallel, RestoreUseSkipNonexistentTargets` through code is when `RestoreBuildInParallel and RestoreUseSkipNonexistentTargets properties set to false`

https://github.com/NuGet/NuGet.Client/compare/release-4.7.3...release-4.8.2

![image](https://user-images.githubusercontent.com/52756182/159761952-eb5fab18-dac1-44e1-9bec-b06d854f00d0.png)

--- a/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
+++ b/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
@@ -378,6 +378,7 @@ Copyright (c) .NET Foundation. All rights reserved.
        Condition=" '%(RestoreGraphProjectInputItems.Extension)' != '.metaproj'
                    AND '%(RestoreGraphProjectInputItems.Extension)' != '.shproj'
                    AND '%(RestoreGraphProjectInputItems.Extension)' != '.vcxitems'
+                   AND '%(RestoreGraphProjectInputItems.Extension)' != '.vdproj'
                    AND '%(RestoreGraphProjectInputItems.Extension)' != '' " />
     </ItemGroup>
 

